Namespaced includes

vkrause opened the pull request:
This is the alternative approach to https://github.com/quotient-im/libQuotient/pull/625, and is source-compatible for CMake consumers.

This still needs to be checked for completeness (ie. are all headers really working in a pure namespaced use now), and needs the opt-out switch in the CMake config files, but it should hopefully be sufficient to decide on which of the approaches to take.

vkrause committed:
Rename lib/ to Quotient/ to match the namespaced include prefix

vkrause committed:
Always install headers into the namespaced include path

The include search path propagated by CMake to consumers is adjusted
accordingly, so this remains source-compatible.

vkrause committed:
Use namespaced includes in generated code

For this to work transparently, add the include install location to the
search path as well.

vkrause committed:
Use namespaced includes in non-generated header files as well

vkrause committed:
Allow consumers to enforce using namespaced includes

Setting QUOTIENT_FORCE_NAMESPACED_INCLUDES before the find_package
call will not add the namespace include dir to the search path.

This isn't really elegant, but allows both retaining full source
compatibility and helps consumers that have clashing include file
names.

vkrause commented:
Tested by including all headers in the namespaced form in consumer code and building with the namespace folder removed from the search path. Also verified that an unmodified Neochat still builds fine after this.

vkrause commented:
@KitsuneRal so how should we proceed with this?

KitsuneRal commented:
Upon a quick glance, I certainly like this approach over #625. I will take a closer look at it next week.

vkrause commented:
> Upon a quick glance, I certainly like this approach over #625. I will take a closer look at it next week.

Any news on this?

KitsuneRal commented:
I tested it out with Quaternion and it actually looks quite good. Can you please resolve the conflicts (I think the only actual one is in `CMakeLists.txt`), and I will merge it.

KitsuneRal commented on CMakeLists.txt:
```diff
@@ -309,11 +308,12 @@ endif()
 target_include_directories(${PROJECT_NAME} PUBLIC
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/lib>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/Quotient>
```
I think this include path should also obey `QUOTIENT_FORCE_NAMESPACED_INCLUDES`, otherwise using libQuotient as a submodule is not coherent with using it as an external library (this is the case of Quaternion: it can use the library both ways). Something like
```suggestion
    $<BUILD_INTERFACE:$<$<NOT:$<BOOL:${QUOTIENT_FORCE_NAMESPACED_INCLUDES}>>:${CMAKE_CURRENT_SOURCE_DIR}/Quotient>
```
That, in turn will cause the library to fail building because quite a few files use the fact that the topmost source directory is in the include path. This is not a good assumption anyway, and these files should be corrected - except quotest and autotests, perhaps. Ideally I'd like `QUOTIENT_FORCE_NAMESPACED_INCLUDES` not to apply to them (not to autotests for sure; quotest is arguable).

vkrause replied:
I've made libQuotient itself buildable with namespaced includes, I still have to do the buildsystem part in a way it doesn't affect the tests though.

KitsuneRal replied:
This is what I got as a result: https://github.com/quotient-im/libQuotient/tree/namespaced-includes

KitsuneRal replied:
...oh, alright, I'm just a bit too late :)

KitsuneRal replied:
...and that being said - you still haven't applied this suggestion (i.e. actually enforcing `QUOTIENT_FORCE_NAMESPACED_INCLUDES` in `BUILD_INTERFACES`) - is it intentional?

vkrause replied:
That's what I meant with "I still have to do the buildsystem part" above :) (I'm at LAS at the moment and have to do this in the breaks).

vkrause replied:
... and should all be there now.

KitsuneRal reviewed (commented):
Ok, I actually have one suggestion, but I'm fine to merge this PR without applying it, and implement that suggestion on my end as a separate PR instead.

vkrause committed:
Make libQuotient itself buildable with namespaced includes

vkrause committed:
Use namespaced includes in test code as well

This is necessary in order to also allow enforcing namespaced includes
in embedded/inline builds of libQuotient and is cleaner to do that way
then to manually mess with include search paths just for the tests.

vkrause committed:
Also honor QUOTIENT_FORCE_NAMESPACED_INCLUDES for inline/embedded builds
